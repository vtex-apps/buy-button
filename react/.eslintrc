{
  "extends": "vtex-react",
  "env": {
    "browser": true,
    "es6": true,
    "jest": true
  },
  "rules": {
    "react/jsx-filename-extension": "off",
    "@typescript-eslint/ban-ts-comment": "off"
  },
  "overrides": [
    {
      "files": ["*.js", "*.jsx"],
      "parser": "babel-eslint",
      "parserOptions": {
        "sourceType": "module"
      }
    }
  ]
}
