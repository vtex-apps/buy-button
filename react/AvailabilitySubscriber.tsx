import React, { useState } from 'react'
import type { FormEvent, ChangeEvent } from 'react'
import { useIntl } from 'react-intl'
import { useMutation } from 'react-apollo'
import { Button, Input } from 'vtex.styleguide'
import { useProduct } from 'vtex.product-context'
import type { Seller } from 'vtex.product-context'

import ADD_TO_AVAILABILITY_SUBSCRIBER_MUTATION from './graphql/addToAvailabilitySubscriberMutation.gql'
import styles from './AvailabilitySubscriber.css'
import { getDefaultSeller } from './utils/sellers'

interface MutationVariables {
  acronym: string
  document: {
    fields: Array<{
      key: string
      value?: string | null
    }>
  }
}
interface Props {
  /* Product's availability */
  available?: boolean
  /* SKU id to subscribe to */
  skuId?: string
}

const isAvailable = (commertialOffer?: Seller['commertialOffer']) => {
  return (
    commertialOffer &&
    (Number.isNaN(+commertialOffer.AvailableQuantity) ||
      commertialOffer.AvailableQuantity > 0)
  )
}

/**
 * Availability Subscriber Component.
 * A form where users can sign up to be alerted
 * when a product becomes available again
 */
function AvailabilitySubscriber(props: Props) {
  const productContext = useProduct()
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [emailError, setEmailError] = useState(false)
  const [didBlurEmail, setDidBlurEmail] = useState(false)

  const [signUp, { loading, error, data }] = useMutation<
    unknown,
    MutationVariables
  >(ADD_TO_AVAILABILITY_SUBSCRIBER_MUTATION)

  const intl = useIntl()

  const seller = getDefaultSeller(productContext.selectedItem?.sellers)

  const available = props.available ?? isAvailable(seller?.commertialOffer)
  const skuId = props.skuId ?? productContext.selectedItem?.itemId

  // Render component only if the product is out of stock
  if (available || !skuId) {
    return null
  }

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()

    const variables: MutationVariables = {
      acronym: 'AS',
      document: {
        fields: [
          {
            key: 'skuId',
            value: skuId,
          },
          {
            key: 'name',
            value: name,
          },
          {
            key: 'email',
            value: email,
          },
          {
            key: 'notificationSend',
            value: 'false',
          },
          {
            key: 'createdAt',
            value: new Date().toISOString(),
          },
          {
            key: 'sendAt',
            value: null,
          },
        ],
      },
    }

    const signUpMutationResult = await signUp({
      variables,
    })

    if (!signUpMutationResult.errors) {
      setName('')
      setEmail('')
    }

    const event = new CustomEvent('message:success', {
      detail: {
        success: true,
        message: intl.formatMessage({
          id: 'store/availability-subscriber.added-message',
        }),
      },
    })

    document.dispatchEvent(event)
  }

  const validateEmail = (newEmail: string) => {
    const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

    setEmailError(!emailRegex.test(newEmail.toLowerCase()))
  }

  const handleEmailChange = (e: ChangeEvent<HTMLInputElement>) => {
    setEmail(e.target.value)
    validateEmail(e.target.value)
  }

  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    setName(e.target.value)
  }

  let emailErrorMessage = ''

  if (didBlurEmail && emailError) {
    emailErrorMessage = intl.formatMessage({
      id: 'store/availability-subscriber.invalid-email',
    })
  }

  const isFormDisabled = name === '' || email === '' || emailError || loading

  return (
    <div className={styles.subscriberContainer}>
      <div className={`${styles.title} t-body mb3`}>
        {intl.formatMessage({ id: 'store/availability-subscriber.title' })}
      </div>
      <div className={`${styles.subscribeLabel} t-small fw3`}>
        {intl.formatMessage({
          id: 'store/availability-subscriber.subscribe-label',
        })}
      </div>
      <form className={`${styles.form} mb4`} onSubmit={e => handleSubmit(e)}>
        <div className={`${styles.content} flex-ns justify-between mt4 mw6`}>
          <div className={`${styles.input} ${styles.inputName} w-100 mr5 mb4`}>
            <Input
              name="name"
              type="text"
              placeholder={intl.formatMessage({
                id: 'store/availability-subscriber.name-placeholder',
              })}
              value={name}
              onChange={handleNameChange}
            />
          </div>
          <div className={`${styles.input} ${styles.inputEmail} w-100 mr5 mb4`}>
            <Input
              name="email"
              type="text"
              placeholder={intl.formatMessage({
                id: 'store/availability-subscriber.email-placeholder',
              })}
              value={email}
              onChange={handleEmailChange}
              onBlur={() => setDidBlurEmail(true)}
              error={didBlurEmail && emailError}
              errorMessage={emailErrorMessage}
            />
          </div>
          <div className={`${styles.submit} flex items-center mb4`}>
            <Button
              type="submit"
              variation="primary"
              size="small"
              disabled={isFormDisabled}
              isLoading={loading}
            >
              {intl.formatMessage({
                id: 'store/availability-subscriber.send-label',
              })}
            </Button>
          </div>
        </div>
        {!error && data && (
          <div className={`${styles.success} t-body c-success`}>
            {intl.formatMessage({
              id: 'store/availability-subscriber.added-message',
            })}
          </div>
        )}
        {error && (
          <div className={`${styles.error} c-danger`}>
            {intl.formatMessage({
              id: 'store/availability-subscriber.error-message',
            })}
          </div>
        )}
      </form>
    </div>
  )
}

export default AvailabilitySubscriber
