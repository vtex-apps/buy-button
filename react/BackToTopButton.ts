import BackToTopButton from './components/BackToTopButton'

export default BackToTopButton
