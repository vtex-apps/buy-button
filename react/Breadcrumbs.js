import Breadcrumbs from './components/Breadcrumbs/index'

export default Breadcrumbs
