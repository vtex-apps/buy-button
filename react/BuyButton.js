import BuyButton from './components/buy-button/src/index'

export default BuyButton
