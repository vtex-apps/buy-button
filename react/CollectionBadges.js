import CollectionBadges from './components/collection-badges/index'

export default CollectionBadges
