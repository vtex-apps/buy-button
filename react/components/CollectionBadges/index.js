import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { CollectionBadgeItem } from './components/CollectionBadgeItem'
import VTEXClasses from './components/CustomClasses'

/**
 * Collection Badges component.
 * Encapsulates and displays a responsive list of Collection Badges.
 */
const CollectionBadges = ({ collectionBadgesText, children }) => (
  <div className={`${VTEXClasses.COLLECTION_BADGES} relative dib h-100`}>
    {children}
    <div className="inline-flex justify-end absolute w-100 bottom-0 left-0">
      {collectionBadgesText.map(collectionBadgeText => (
        <CollectionBadgeItem key={collectionBadgeText}>
          {collectionBadgeText}
        </CollectionBadgeItem>
      ))}
    </div>
  </div>
)

CollectionBadges.propTypes = {
  /** Array of collection badges text */
  collectionBadgesText: PropTypes.array.isRequired,
  /** Children component that should be render inside the collection badge item */
  children: PropTypes.node.isRequired,
}

CollectionBadges.defaultProps = {
  collectionBadgesText: [],
}

export default CollectionBadges
