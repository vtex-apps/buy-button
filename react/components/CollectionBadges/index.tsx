import React from 'react'

import { CollectionBadgeItem } from './components/CollectionBadgeItem'
import styles from './styles.css'

type OwnProps = {
  collectionBadgesText: any[]
  children: React.ReactNode
}

// @ts-expect-error ts-migrate(2456) FIXME: Type alias 'Props' circularly references itself.
type Props = OwnProps & typeof CollectionBadges.defaultProps

/**
 * Collection Badges component.
 * Encapsulates and displays a responsive list of Collection Badges.
 */
// @ts-expect-error ts-migrate(7022) FIXME: 'CollectionBadges' implicitly has type 'any' becau... Remove this comment to see the full error message
const CollectionBadges = ({ collectionBadgesText, children }: Props) => (
  <div className={`${styles.collectionContainer} relative dib w-100`}>
    {children}
    <div className="inline-flex justify-end absolute w-100 bottom-0 left-0">
      {collectionBadgesText.map((collectionBadgeText: any, index: any) => (
        <CollectionBadgeItem key={collectionBadgeText + index}>
          {collectionBadgeText}
        </CollectionBadgeItem>
      ))}
    </div>
  </div>
)

CollectionBadges.defaultProps = {
  collectionBadgesText: [],
}

export default CollectionBadges
