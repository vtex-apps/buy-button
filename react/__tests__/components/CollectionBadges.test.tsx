import React from 'react'
import { render } from '@vtex/test-tools/react'

import CollectionBadges from '../../CollectionBadges'

describe('<CollectionBadges /> component', () => {
  function renderComponent(customProps = {}) {
    const component = <CollectionBadges {...customProps}>Test</CollectionBadges>

    return render(component)
  }

  it('should be able to mount and not break', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toBeTruthy()
  })

  it('should match snapshot without badges', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot with badges', () => {
    const { asFragment } = renderComponent({
      collectionBadgesText: ['badge1', 'badge2'],
    })

    expect(asFragment()).toMatchSnapshot()
  })
})
