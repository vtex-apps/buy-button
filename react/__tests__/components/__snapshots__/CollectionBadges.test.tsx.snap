// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollectionBadges /> component should match snapshot with badges 1`] = `
<DocumentFragment>
  <div
    class="collectionContainer relative dib w-100"
  >
    Test
    <div
      class="inline-flex justify-end absolute w-100 bottom-0 left-0"
    >
      <div
        class="item mh1 pa2 bg-blue white tc"
      >
        badge1
      </div>
      <div
        class="item mh1 pa2 bg-blue white tc"
      >
        badge2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<CollectionBadges /> component should match snapshot without badges 1`] = `
<DocumentFragment>
  <div
    class="collectionContainer relative dib w-100"
  >
    Test
    <div
      class="inline-flex justify-end absolute w-100 bottom-0 left-0"
    />
  </div>
</DocumentFragment>
`;
